src/garage/_dtypes.py

```diff
@@ -559,3 +559,214 @@ def output_space(self):
 
         """
         return self._output_space
+
+
+class SampleBatch(
+        collections.namedtuple('SampleBatch', [
+            'observations', 'actions', 'rewards', 'next_observations',
+            'terminals', 'batch_size'
+        ])):
+    # pylint: disable=missing-param-doc, missing-type-doc
+    """A tuple representing a batch of samples.
+
+    Attributes:
+        observations (numpy.ndarray): Non-flattened array of observations.
+            Typically has shape (batch_size, S^*) (the unflattened state space
+            of the current environment).
+        actions (numpy.ndarray): Non-flattened array of actions. Should
+            have shape (batch_size, S^*) (the unflattened action space of the
+            current environment).
+        rewards (numpy.ndarray): Array of rewards of shape (batch_size,) (1D
+            array of length batch_size).
+        next_observation (numpy.ndarray): Non-flattened array of next
+            observations. Has shape (batch_size, S^*). next_observations[i] was
+            observed by the agent after taking actions[i].
+        terminals (numpy.ndarray): A boolean numpy array of shape
+            shape (batch_size,) containing the termination signals for all
+            transitions in this batch.
+        batch_size (int): The size of the batch
+
+    Raises:
+        ValueError: If any of the above attributes do not conform to their
+            prescribed types and shapes.
+    """
+    __slots__ = ()
+
+    def __new__(cls, observations, actions, rewards, next_observations,
```

Comment:
why not take env_spec so you can verify validate the observations and actions? we should either do this in all dtypes or none of them.